### ISSUE TITLE
Optimizer clipvalue and clipnorm not working in Tensorflow 2.0

### ISSUE THREAD

**Issue by bloc97** [tensorflow/tensorflow#33929]
**System information**
- Python 3 Google Compute Engine backend (GPU)

**Describe the current behavior**
clipvalue and clipnorm in Optimizers does nothing!

**Describe the expected behavior**
By setting clipvalue=0 or clipnorm=0 no training should occur (gradients should be 0!), but the network still trains, and if using a large learning rate, loss goes to nan.

**Code to reproduce the issue**
![image](https://user-images.githubusercontent.com/567732/68066124-4d01b000-fd09-11e9-8e41-51ece3684869.png)
**Gradient is clearly not zero since the network is getting modified at each iteration.**

![image](https://user-images.githubusercontent.com/567732/68066145-7589aa00-fd09-11e9-8c94-1ae4bcf33e45.png)
**Sanity check by setting lr=0**
**No training occurs when lr=0, as expected.**

**Comment by ravikyram** [comment]
Looks like code is incomplete.Please provide a simple standalone code to reproduce the issue in our environment, then it is easy for localizing the issue faster. Thanks!

**Comment by ravikyram** [comment]
@bloc97 

Any update on this issue please. Thanks!

**Comment by bloc97** [comment]
I didn't think that standalone code would be needed as this unexpected behavior should be instantly caught by a unit test. But here you go:

https://colab.research.google.com/drive/16gxaqBuY7NRYZaDqeV26yH7caLChTLY6

**Comment by bloc97** [comment]
Just for completeness' sake, here's the same code but on Tensorflow 1.x:
https://colab.research.google.com/drive/1BylY2e43DLZRBkOdIDGJ_rBPDlbGo-Vm

**Comment by ravikyram** [comment]
@bloc97 
I tried in colab with TF 2.0 for learning rate 0 and 100 and i am able to see loss in output. However loss goes to nan if we use very large value `learning rate: 1e10`. Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/af23720e98b0f655f794773c30cd486c/untitled371.ipynb). However with TF 1.15 version i am able to see loss for learning rate 0, 100 and 1e10 .Thanks!

**Comment by rmothukuru** [comment]
@bloc97,
Agree that  `learning rate: 1e10` is resulting in `NaN` but can you please let me know why do you want to use such a Huge Value for Learning Rate. Thanks!

**Comment by bloc97** [comment]
Many research papers using high learning rate regimes will diverge if gradient clipping does not work. I simply provided a small example that shows the issue.
For example, in VDSR the authors use a learning rate of 0.1 with gradient clipping of 0.001. Since the network is extremely deep (20+ layers), it will diverge even on learning rates of 0.1 or 0.01 without gradient clipping.

I discovered this issue when trying to run code that worked on TF 1.5, but as we are migrating to TF 2.0, it doesn't work anymore.

**Comment by bloc97** [comment]
Any updates on this matter? Or is there any workaround (manually clip the gradients before the update pass)?

**Comment by tomerk** [comment]
This should be fixed for 2.2 (If you're not using a distribution strategy). We're working on a longer-term change for optimizers that should fix this more fundamentally.

**Comment by kriskorrel-cw** [comment]
I assume that this is the fix: https://github.com/tensorflow/tensorflow/commit/69da929ad4d5ba605507efa1f52b382a55b6a969
May I ask whether this fix is only relevant for eager execution or also graph execution?
In other words: For graph execution, did clipnorm already work correctly?

For me the situation is that I was able to train my models in TF2.0 with graph execution. Then I upgraded to 2.1 and enabled eager execution (and made some other changes), and since, my models cannot train properly anymore. So knowing whether this only relates to eager execution would allow me to pinpoint the problem.
When upgrading to 2.2 (nightly) the problem seems to be fixed.

**Comment by tomerk** [comment]
Yes, norm clipping still worked correctly when eager execution was disabled in 2.0 and 2.1.

**Comment by kriskorrel-cw** [comment]
> Yes, norm clipping still worked correctly when eager execution was disabled in 2.0 and 2.1.

Thanks for clearing that up.

**Comment by jlherren** [comment]
Oh man, I've just been debugging my NaN losses for over an hour now. How could this even go unnoticed...

Will this be fixed for 2.1.x as well or is 2.2 the only way to get this working again? I can't disable eager mode because then I lose the cuDNN implementation of LSTMs.

**Comment by tomerk** [comment]
Hi Goldie, someone wants to know if a bug fix in 2.2 can be backported to
2.1.x. What's our policy on minor release versions?

Best,
Tomer

**Comment by jvishnuvardhan** [comment]
@tomerk Our policy about back-porting the code to previous versions is mentioned by  @mihaimaruseac [Here](https://github.com/tensorflow/tensorflow/issues/29624#issuecomment-610601027). Thanks!

**Comment by karlhjm** [comment]
> This should be fixed for 2.2 (If you're not using a distribution strategy). We're working on a longer-term change for optimizers that should fix this more fundamentally.

Does it work for distribution strategy in 2.2 now? I get the same error when I use MultiWorkerMirroredStrategy in 2.2.0-dev20200508.

**Comment by zaccharieramzi** [comment]
@tomerk can you elaborate on why one shouldn't use gradient clipping with a distrubtion strategy?

**Comment by tomerk** [comment]
@karlhjm no, we still disable it in 2.2 w/ distribution strategies enabled.

@zaccharieramzi yes, happy to elaborate:
There's two possible places to clip when you have distribution strategies enabled:
- before gradients get aggregated (usually wrong)
- after gradients get aggregated (usually right & what people expect)

We want it working w/ the second case (clipping after gradients are aggregated).
The issue is the optimizers are written with clipping happening in the code before aggregation does.

We looked into changing this, but it would have required either:
1. api changes that break existing users of optimizer `apply_gradients`/other non-minimize methods
2. changing the signatures of methods optimizer implementers need to implement, breaking existing custom optimizers

So rather than:
* quietly doing clipping in the wrong place
* increasing churn & breaking existing users or existing custom optimizers just for this individual feature

We instead decided to leave this disabled for now. We'll roll support for this into a larger optimizer refactoring that solves a larger set of issues.
(RFC for that is at https://github.com/tensorflow/community/pull/234)

**Comment by bjourne** [comment]
Short of reducing the learning rate is there any way to workaround this issue?

**Comment by w4nderlust** [comment]
> Short of reducing the learning rate is there any way to workaround this issue?

In Ludwig I have implemented a workaround: A wrapper class that wraps the optimizer and clips the gradients before applying them: https://github.com/uber/ludwig/blob/tf2_porting/ludwig/models/modules/optimization_modules.py
It's still WIP but seems to work fine.

**Comment by Saduf2019** [comment]
@bloc97 
I have tried the code on tf 2.3 and do not get any nan values as output, can you please refer to this gist and confirm if it resolves the issue.
Please find the [gist here](https://colab.research.google.com/gist/Saduf2019/0cc40cbbc833655d43a7b5567d5f212e/untitled321.ipynb)

**Comment by bloc97** [comment]
The issue has been fixed for me on Tensorflow v2.3.0 on google colab. Thanks!

**Comment by tomerk** [comment]
Yes, the optimizer RFC I mentioned above is now complete and this should be resolved :)

**Comment by w4nderlust** [comment]
One question seems still open though. TF provides `tf.clip_by_norm`, `tf.clip_by_value` and `tf.clip_by_global_norm`, while the current implementation only supports the first two, but not the global norm one. Should i open a new issue to have it included?

**Comment by tanzhenyu** [comment]
> One question seems still open though. TF provides `tf.clip_by_norm`, `tf.clip_by_value` and `tf.clip_by_global_norm`, while the current implementation only supports the first two, but not the global norm one. Should i open a new issue to have it included?

THere is a `clip_globalnorm` with keras optimizer in nightly

**Comment by jhanilesh96** [comment]
Still not resolved in conda tf 2.3.0
